### ISSUE TITLE
confusing application state example

### ISSUE THREAD

**Issue by pythoneer** [actix/actix-web#531]
I find the example for `Application State` both in the [guide](https://actix.rs/docs/application/) and the actual [example](https://github.com/actix/examples/blob/master/state/src/main.rs) really confusing. Both for newcomers and intermediate users (not necessarily in a Rust way)

[This issue](https://github.com/actix/actix-web/issues/459) is a good example of how this confusion is coming across. In this particular case the "user" was actually capable of investigating the problem by them self but just lacked some insides in Rust and Actix-Web. 

The main "Problem" here is, that  we have multiple instances of the `Application State` for every worker thread (reactor?) that is created (as stated in the Guide and example). So we have a copy of it in every thread that is handling a request. That by itself is not problematic if you're familiar with it. The problem – as i see it – is the example by itself. If you have no idea of how the parts of actix-web work together you could get the impression that the example would have a single shared mutable state for all (threads ... or maybe you don't even know that there are threads altogether) invocations of a handler.

Running the example you get weird behavior (sometimes the counter goes up, sometimes it goes back and then jumps up again ... etc) because the handler uses a different instance of the `Application State` for some invocations of the handler. I find this very confusing and i cant really think of a useful way to have such a behavior. 

I would imagine to have the example written in such a way that introduce something like a version with `Arc` | `RwLock` etc. to have an experience similar to other web frameworks (not necessarily Rust ones). 

I had to read the Guide twice (at the same time wondering about the function signature) and scratching my head what the author really want to achieve with the `Cell` ... and after a few moments it clicked. But the reason was that i am familiar with the Rust type system and i know what to do. I still don't see the usefulness of counting the invocation of the handler for every instance for the application state in several threads thus having the count "jump around". 

what do you think, maybe i just have misinterpreted the example and guide but i can clearly see how this can be a problem for many people.

**Comment by DoumanAsh** [comment]
If you read carefully you notice note in guide itself:

>Note: http server accepts an application factory rather than an application instance. Http server constructs an application instance for each thread, thus application state must be constructed multiple times. If you want to share state between different threads, a shared object should be used, e.g. Arc. Application state does not need to be Send and Sync, but the application factory must be Send + Sync.

PRs to improve documentation and examples are welcome though

**Comment by pythoneer** [comment]
I am very aware of the note (as stated in my original post). I would happily provide a PR for this – i am just wondering if "we" can agree on the problem and a way to fix it. It is wonderful that there is a hint in the comment, but why have the example be in a way that is confusing? I am just searching consensus to justify such a change. Maybe i am wrong and making the example worse. 

From my point of view, the example suggest that the handler is counting the invocations  and reports them back to the (webservice) caller [if i wget this service 10 times i expect a result like (0,1,2,3,4,5,6,7,8,9) for every consecutive invocation]– where in fact what the caller gets back depends on which thread the handler is currently running on and can get back a count of (0,1,2,3,0,1,2,4,5,6,3,4 .. ) which i think is not the lesson the example and the guide wants to teach here. 

We could still have this in the guide (as a way to show how to not do this kind of behavior) and show a way to "correct" this (with e.g an Arc)  and have the example with the "correct" solution and a hint back to the guide. What is the use of a copied counter in every thread anyway?

What do you think?

**Comment by DoumanAsh** [comment]
> but why have the example be in a way that is confusing?

It is subjective matter. To me it is pretty clear that example aims to be as simple as possible.

There is no correct or right way to define state.
It is always up to user.

But to summarize: I don't feel strongly about it, but I understand your point and I would welcome PR to use Arc for state.

**Comment by fdns** [comment]
Hello, I was learning to use actix-web, and I got to this exact confusion. I actually first did something like the [example](https://github.com/actix/examples/pull/54) (Arc<Mutex<>>), but it doesn't work as intended, because server::new gets called for every worker separately, creating a new independent Arc (and State object).

For this to work, I did something like:
```
let state = Arc::new(RwLock::new(api::state()));
server::new(move || { vec![App::with_state(state.read().unwrap().clone()).resource("/", ....)]
});
```
I'm not the best to fix the documentation, but probably a lot of people will shoot itself in the foot trying the proposed solution.
